### PR TITLE
Support auto transitions events

### PR DESCRIPTION
**Pull request by pelegdvir**
* New classmethod - `all_states` - list all allowed states in the state machine
* New classmethod - `available_transitions` - list all allowed next possible events
* New feature - `autotransition` - events that would automatically get created if possible (after every factory call, try to call any valid autotransition event)
* Assumption: only one (or zero) `autotransition` can follow any state. No other event may follow.
* New classmethod - `has_only_valid_transiations` - validates the above rule

### CODE REVIEW

File: microcosm_eventsource/event_types.py [hunk 1]
```diff
@@ -33,17 +33,20 @@ class EventTypeInfo:
...
+    def __init__(self, follows=None, accumulate=None, restarting=False, requires=(), autotransition=False):
         """
-        :param follows:    an instance of the event mini-grammar
-        :param accumulate: whether the event type should accumulating state
-        :param restarting: whether the event type may restart a (new) version
+        :param follows:         an instance of the event mini-grammar
+        :param accumulate:      whether the event type should accumulating state
+        :param restarting:      whether the event type may restart a (new) version
+        :param requires:        validate type-specific required fields
+        :param autotransition:  whether this event should be autogenereated when possible
```
Comment (jessemyers) [inline microcosm_eventsource/event_types.py]: I would use `auto_transition`

File: microcosm_eventsource/event_types.py [hunk 2]
```diff
@@ -140,6 +169,47 @@ def next_version(self, version):
...
         return version
 
+    @classmethod
+    def all_states(cls):
+        """
+        All states that the state machine can reach
+
+        """
+        # Prefer to use frozenset and not set to represent states in this context
+        # It will allow us easly ignore for duplications
```
Comment (jessemyers) [inline microcosm_eventsource/event_types.py]: I don't follow what you mean on L179.

File: microcosm_eventsource/event_types.py [hunk 3]
```diff
@@ -140,6 +169,47 @@ def next_version(self, version):
...
+        intial_states = [frozenset((event_type,)) for event_type in cls if event_type.is_initial]
+        next_states = intial_states
+        known_states = set(intial_states)
+
+        while True:
```
[the remark this answers is not in the document]
Comment (jessemyers) [reply microcosm_eventsource/event_types.py]: Can we check somewhere whether an iteration fails to change the traversal? (And abort with an exception?)

File: microcosm_eventsource/event_types.py [hunk 4]
```diff
@@ -140,6 +169,47 @@ def next_version(self, version):
...
+            for state in current_states:
+                for event_type in cls.available_transitions(state):
+                    new_state = frozenset(event_type.accumulate_state(state))
+                    if new_state in known_states:
+                        continue
+                    next_states.append(new_state)
+                    known_states.add(new_state)
+
+    @classmethod
+    def has_only_valid_transiations(cls):
```
[the remark this answers is not in the document]
Comment (jessemyers) [reply microcosm_eventsource/event_types.py]: Your spelling is off here: `s/transiations/transitions/`

File: microcosm_eventsource/factory.py [hunk 5]
```diff
@@ -74,6 +75,19 @@ def create_transition(self, event_info, **kwargs):
         self.process_state_transition(event_info)
         self.create_event(event_info, **kwargs)
 
+    def create_autotransition_event(self, ns, sns_producer, parent, **kwargs):
+        """
+        Creates the next autotransition event if exist
+
+        """
+        auto_transition_events = [
```
Comment (jessemyers) [inline microcosm_eventsource/factory.py]: See, here you got the naming right: `auto_transition`

File: microcosm_eventsource/tests/test_crud.py [hunk 6]
```diff
@@ -299,3 +299,42 @@ def test_transition_to_revised(self):
...
+                call(
+                    media_type="application/vnd.globality.pubsub._.created.task_event.completed",
+                    uri="http://localhost/api/v1/task_event/{}".format(data["id"]),
+                ),
+                call(
+                    media_type="application/vnd.globality.pubsub._.created.task_event.ended",
+                    uri="http://localhost/api/v1/task_event/{}".format(ended_task_event.id),
+                ),
+            ],
+            any_order=True,
```
[the remark this answers is not in the document]
Comment (jessemyers) [reply microcosm_eventsource/tests/test_crud.py]: It's new to me, but it seems useful.

File: microcosm_eventsource/tests/test_event_types.py [hunk 7]
```diff
@@ -191,3 +206,44 @@ def test_completed_is_terminal():
 
     # may not be cancelled
     assert_that(TaskEventType.CANCELED.may_transition(state), is_(equal_to(False)))
+
+
+def test_all_states():
+    """
+    Find all allowd states.
```
Comment (jessemyers) [inline microcosm_eventsource/tests/test_event_types.py]: spelling: allowed

File: microcosm_eventsource/event_types.py [hunk 8]
```diff
@@ -140,6 +169,47 @@ def next_version(self, version):
...
+        * only one (or zero) autotransition event can follow any state. No other event may follow.
+
+        """
+        for state in cls.all_states():
+            auto_transition_events = [
+                event_type for event_type in cls.autotransition_events()
+                if event_type.may_transition(state)
+            ]
+            if auto_transition_events and len(cls.available_transitions(state)) > 1:
+                return False
```
[the remark this answers is not in the document]
Comment (saivickna) [reply microcosm_eventsource/event_types.py]: not sure if its needed but do we need to check for loops? A -> B -> C -> A?

File: microcosm_eventsource/tests/test_crud.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saivickna) [reply microcosm_eventsource/tests/test_crud.py]: is this order actually indeterminate?

File: microcosm_eventsource/event_types.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pelegdvir) [reply microcosm_eventsource/event_types.py]: No - we don't worry about cycles / no changes. Just from huge graphs.
I think that generator is actually the solution to this. The responsibility to break would move to the caller.

File: microcosm_eventsource/event_types.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pelegdvir) [reply microcosm_eventsource/event_types.py]: No - loops are legit

**Pull request merged**